### PR TITLE
Fix balance-leader won't work when placement rule enabled

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?
When placement rule enabled, balance-leader won't work.

### What is changed and how it works?
The region copy logic in ruleLeaderFilter's Target function is wrong. This request fix this bug.

### Check List

Tests

- Unit test

Related changes

- Need to cherry-pick to the release branch

### Release note

- Fix the bug that balance-leader won't work when placement rule enabled (imported from v4.0.3)

### CODE REVIEW

File: server/schedule/filter/filters.go [hunk 1]
```diff
@@ -461,9 +464,18 @@ func (f *ruleLeaderFitFilter) Source(opt opt.Options, store *core.StoreInfo) boo
...
-	region := f.region.Clone(core.WithReplacePeerStore(f.oldLeaderStoreID, store.GetID()), core.WithLeader(&oldLeaderPeerReplica))
+	var targetPeer *metapb.Peer
+	for _, peer := range f.region.GetPeers() {
+		if peer.StoreId == store.GetID() {
+			targetPeer = peer
+			break
+		}
+	}
+	if targetPeer == nil {
+		log.Info("ruleLeaderFitFilter could't find peer on target Store", zap.Uint64("newStoreID", store.GetID()))
```
Comment (rleungx) [inline server/schedule/filter/filters.go:475]: Will the log be noisy?

File: server/schedule/filter/filters.go [hunk 2]
```diff
@@ -461,9 +464,18 @@ func (f *ruleLeaderFitFilter) Source(opt opt.Options, store *core.StoreInfo) boo
 }
 
 func (f *ruleLeaderFitFilter) Target(opt opt.Options, store *core.StoreInfo) bool {
-	oldLeaderPeerReplica := *f.region.GetLeader()
-	oldLeaderPeerReplica.StoreId = store.GetID()
-	region := f.region.Clone(core.WithReplacePeerStore(f.oldLeaderStoreID, store.GetID()), core.WithLeader(&oldLeaderPeerReplica))
+	var targetPeer *metapb.Peer
+	for _, peer := range f.region.GetPeers() {
```
Comment (disksing) [inline server/schedule/filter/filters.go:468]: f.region.GetStorePeer()

File: server/schedule/filter/filters.go (hunk 1 above)
Comment (disksing) [reply server/schedule/filter/filters.go:475]: I think normally it should not happen. Maybe `log.Warn` here.

File: server/schedule/filter/filters.go (hunk 2 above)
Comment (Yisaer) [reply server/schedule/filter/filters.go]: updated.

File: server/schedule/filter/filters.go (hunk 1 above)
Comment (Yisaer) [reply server/schedule/filter/filters.go]: updated.

File: server/schedule/filter/filters.go [hunk 3]
```diff
@@ -461,9 +464,13 @@ func (f *ruleLeaderFitFilter) Source(opt opt.Options, store *core.StoreInfo) boo
 }
 
 func (f *ruleLeaderFitFilter) Target(opt opt.Options, store *core.StoreInfo) bool {
-	oldLeaderPeerReplica := *f.region.GetLeader()
-	oldLeaderPeerReplica.StoreId = store.GetID()
-	region := f.region.Clone(core.WithReplacePeerStore(f.oldLeaderStoreID, store.GetID()), core.WithLeader(&oldLeaderPeerReplica))
+	var targetPeer *metapb.Peer
+	targetPeer = f.region.GetStorePeer(store.GetID())
```
Comment (disksing) [inline server/schedule/filter/filters.go:468]:
```suggestion
	targetPeer := f.region.GetStorePeer(store.GetID())
```

File: server/schedule/filter/filters.go [hunk 4]
```diff
@@ -467,7 +467,7 @@ func (f *ruleLeaderFitFilter) Target(opt opt.Options, store *core.StoreInfo) boo
 	var targetPeer *metapb.Peer
 	targetPeer = f.region.GetStorePeer(store.GetID())
 	if targetPeer == nil {
-		log.Warn("ruleLeaderFitFilter could't find peer on target Store", zap.Uint64("newStoreID", store.GetID()))
+		log.Warn("ruleLeaderFitFilter could't find peer on target Store", zap.Uint64("target-store", store.GetID()))
```
Comment (rleungx) [inline server/schedule/filter/filters.go:470]:
```suggestion
		log.Warn("ruleLeaderFitFilter couldn't find peer on target Store", zap.Uint64("target-store", store.GetID()))
```

**Pull request merged**